### PR TITLE
Come to terms with identification

### PR DESCRIPTION
**Pull request by jcorbin**
Some cleanups and a new test plus better containment of identify-related mutation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,18 +11,18 @@ var server = new TChannel({host: '127.0.0.1', port: 4040});
 var client = new TChannel({host: '127.0.0.1', port: 4041});
 
 // normal response
-server.register('func 1', function (arg1, arg2, peerInfo, cb) {
-    console.log('func 1 responding immediately 1:' + arg1.toString() + ' 2:' + arg2.toString());
+server.register('func 1', function (reqHead, reqBody, peerInfo, cb) {
```
Comment (Raynos) [inline README.md]: :+1: for better names.

File: package.json [hunk 2]
```diff
@@ -24,9 +24,11 @@
     "metrics": "^0.1.8",
     "opn": "^1.0.1",
     "pre-commit": "0.0.9",
+    "run-parallel": "^1.0.0",
```
Comment (Raynos) [inline package.json]: thanks :)

File: parser.js [hunk 3]
```diff
@@ -20,29 +20,32 @@
...
-	read_arg3len: 6,
-	read_csum: 7,
-	read_arg1: 8,
-	read_arg2: 9,
-	read_arg3: 10,
-	error: 255
-};
+/* jshint camelcase:false */
+
+var types = {};
```
Comment (Raynos) [inline parser.js]: good idea.

File: index.js [hunk 4]
```diff
@@ -140,17 +153,18 @@ TChannel.prototype.addPeer = function (name, connection) {
...
 		self.removePeer(name, connection);
 	});
 	connection.on('socketClose', function (conn, err) {
 		self.emit('socketClose', conn, err);
 	});
-	connection.remoteName = name;
+	return this.setPeer(name, connection);
 };
 
+/* jshint maxparams:5 */
```
Comment (Raynos) [inline index.js]: jshint cleanup is :+1:

File: index.js [hunk 5]
```diff
@@ -84,6 +84,10 @@ TChannel.prototype.register = function (op, callback) {
 };
 
 TChannel.prototype.setPeer = function (name, conn) {
+	if (name === this.name) {
+		throw new Error('cowardly refusing to set self peer');
```
Comment (jwolski) [inline index.js]: the `cowardly` message is weird. it's cute. but remove it ;).

File: index.js [hunk 6]
```diff
@@ -163,18 +177,17 @@ TChannel.prototype.send = function (options, arg1, arg2, arg3, callback) {
 
 	var peer = this.getPeer(dest);
 	if (!peer) {
-		this.addPeer(dest, this.makeOutConnection(dest));
-		peer = this.getPeer(dest);
+		peer = this.addPeer(dest);
```
Comment (jwolski) [inline index.js]: good one

**Comment by jwolski** [comment]
make sure to run benchmarks on all this jazz

**Comment by jcorbin** [comment]
Dropped the cowards.

Tell me more about this benchmarking, it sounds like it could / should be more automated enforcement / checking?

**Comment by mranney** [comment]
Run bench_server.js in a window, then in another window, run multi_bench.js. This will exercise various degrees of pipelining and request size. Make sure to have your laptop plugged in if you are on a laptop.

**Comment by mranney** [comment]
And yes, doing automated checking / enforcement for performance regressions would be amazing.

**Comment by Raynos** [comment]
lgtm.

**Comment by jcorbin** [comment]
Okay, based on #31's work, this branch does not regress performance:
```
+ node /Users/joshuacorbin/Uber/tchannel/benchmarks/compare.js /Users/joshuacorbin/Uber/tchannel/benchmarks/master.json /Users/joshuacorbin/Uber/tchannel/benchmarks/wat_ident.json
GET large str, 1/5     rate: hi-diff:   -178.04 ( -1.5%)
GET large str, 200/5   rate: hi-diff:   1753.30 (  5.6%)
GET large str, 20000/5 rate: hi-diff:   -767.27 ( -2.0%)
GET large str, 50/5    rate: hi-diff:   2088.06 (  6.6%)
GET small str, 1/5     rate: hi-diff:    -78.97 ( -0.5%)
GET small str, 200/5   rate: hi-diff:    822.40 (  1.6%)
GET small str, 20000/5 rate: hi-diff:   3158.81 (  6.3%)
GET small str, 50/5    rate: hi-diff:   -265.70 ( -0.5%)
PING, 1/5              rate: hi-diff:    292.08 (  1.9%)
PING, 200/5            rate: hi-diff:   -158.25 ( -0.3%)
PING, 20000/5          rate: hi-diff:   -485.63 ( -0.8%)
PING, 50/5             rate: hi-diff:   1168.87 (  2.2%)
SET large buf, 1/5     rate: hi-diff:    268.42 (  2.3%)
SET large buf, 200/5   rate: hi-diff:   1171.31 (  3.7%)
SET large buf, 20000/5 rate: hi-diff:   1030.12 (  3.9%)
SET large buf, 50/5    rate: hi-diff:   1760.54 (  5.5%)
SET large str, 1/5     rate: hi-diff:    130.33 (  1.1%)
SET large str, 200/5   rate: hi-diff:   1316.08 (  4.1%)
SET large str, 20000/5 rate: hi-diff:   1020.25 (  3.9%)
SET large str, 50/5    rate: hi-diff:   1693.26 (  5.3%)
SET small buf, 1/5     rate: hi-diff:    196.08 (  1.3%)
SET small buf, 200/5   rate: hi-diff:   2335.27 (  4.8%)
SET small buf, 20000/5 rate: hi-diff:   2022.44 (  4.4%)
SET small buf, 50/5    rate: hi-diff:   2745.32 (  5.8%)
SET small str, 1/5     rate: hi-diff:    307.48 (  2.1%)
SET small str, 200/5   rate: hi-diff:   1344.83 (  2.8%)
SET small str, 20000/5 rate: hi-diff:   3148.05 (  6.9%)
SET small str, 50/5    rate: hi-diff:   2544.31 (  5.4%)
```

if anything that's a specious speed up

**Comment by Raynos** [comment]
#suchwin.

**Pull request merged**